File: src/main/java/me/geek/tom/serverutils/sethome/HomesComponentImpl.java
package me.geek.tom.serverutils.sethome;

import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
import me.geek.tom.serverutils.TomsServerUtils;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.text.LiteralText;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.registry.RegistryKey;
import net.minecraft.world.World;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HomesComponentImpl implements HomesComponent {

    private static final DynamicCommandExceptionType HOME_EXISTS = new DynamicCommandExceptionType(name -> new LiteralText("You already have a home called: " + name));
    public static final DynamicCommandExceptionType HOME_NOT_FOUND = new DynamicCommandExceptionType(name -> new LiteralText("You do not have a home called: " + name));
    private static final DynamicCommandExceptionType TOO_MANY_HOMES = new DynamicCommandExceptionType(max -> new LiteralText("Could not create a new home, you are at the limit of " + max + " homes!"));

    private final List<Home> homes = new ArrayList<>();
    private final Object2ObjectOpenHashMap<String, Home> homesByName = new Object2ObjectOpenHashMap<>();
    private final Object2ObjectOpenHashMap<Identifier, List<Home>> homesByWorld = new Object2ObjectOpenHashMap<>();

    // Used as we need to implement EntityComponentFactory<PlayerComponent<? extends HomesComponent>, PlayerEntity>
    // with a constructor.
    @SuppressWarnings({"RedundantSuppression", "unused"})
    public HomesComponentImpl(PlayerEntity player) { }

    @Override
    public @Nullable Home getByName(@NotNull String name) {
        return this.homesByName.get(name);
    }

    @Override
    public @NotNull List<Home> getAllInDimension(@NotNull RegistryKey<World> dimension) {
        return Collections.unmodifiableList(this.homesByWorld.computeIfAbsent(dimension.getValue(), __ -> new ArrayList<>()));
    }

    @Override
    public @NotNull List<Home> getAllHomes() {
        return Collections.unmodifiableList(this.homes);
    }

    @Override
    public @NotNull Home createNewHome(@NotNull String name, @NotNull RegistryKey<World> dimension, @NotNull BlockPos pos) throws CommandSyntaxException {
        if (this.homesByName.containsKey(name)) throw HOME_EXISTS.create(name);
        int maxHomes = TomsServerUtils.homesConfig.getMaxHomeAmount();
        if (maxHomes != -1 && this.homes.size() == maxHomes) throw TOO_MANY_HOMES.create(maxHomes);

        Home home = new Home(name, dimension, pos);
        this.homes.add(home);
        // Update maps for fast lookup.
        this.homesByName.put(home.getName(), home);
        this.homesByWorld.computeIfAbsent(home.getDimension().getValue(), __ -> new ArrayList<>()).add(home);
        return home;
    }

    @Override
    public void removeHome(@NotNull Home home) {
        this.homesByName.remove(home.getName());
        this.homesByWorld.remove(home.getDimension().getValue());
        this.homes.remove(home);
    }

    @Override
    public void removeHomeByName(@NotNull String name) throws CommandSyntaxException {
        if (!this.homesByName.containsKey(name)) {
            throw HOME_NOT_FOUND.create(name);
        }
        this.removeHome(this.homesByName.get(name));
    }

    @Override
    public void removeHomesInDimension(@NotNull RegistryKey<World> dimension) {
        this.homesByWorld.computeIfAbsent(dimension.getValue(), __ -> new ArrayList<>()).forEach(this::removeHome);
    }

    @Override
    public void readFromNbt(@NotNull CompoundTag tag) {
        this.homes.clear();
        this.homesByWorld.clear();
        this.homesByName.clear();

        ListTag homes = tag.getList("Homes", 10);
        homes.stream().map(t -> (CompoundTag)t).map(Home::read).forEach(this.homes::add);

        // Compute the maps for easier lookup
        this.homes.forEach(h -> this.homesByName.put(h.getName(), h));
        this.homes.forEach(h -> this.homesByWorld.computeIfAbsent(h.getDimension().getValue(), __ -> new ArrayList<>()).add(h));
    }

    @Override
    public void writeToNbt(@NotNull CompoundTag tag) {
        ListTag homes = new ListTag();
        this.homes.stream().map(home -> home.write(new CompoundTag())).forEach(homes::add);
        tag.put("Homes", homes);
    }
}

File: src/main/kotlin/homes/HomesComponentImpl.kt
package me.geek.tom.serverutils.homes

import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import me.geek.tom.serverutils.homesConfig
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtList
import net.minecraft.text.TranslatableText
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.RegistryKey
import net.minecraft.world.World
import java.util.*
import java.util.function.Consumer

// Used as we need to implement EntityComponentFactory<PlayerComponent<? extends HomesComponent>, PlayerEntity>
// with a constructor.
class HomesComponentImpl(player: PlayerEntity) : HomesComponent {

    private val homes: MutableList<Home> = ArrayList()
    private val homesByName = Object2ObjectOpenHashMap<String, Home>()
    private val homesByWorld = Object2ObjectOpenHashMap<Identifier, MutableList<Home>>()

    override fun getByName(name: String): Home? {
        return homesByName[name]
    }

    override fun getAllInDimension(dimension: RegistryKey<World>): List<Home> {
        return Collections.unmodifiableList(homesByWorld.computeIfAbsent(dimension.value) { ArrayList() })
    }

    override val allHomes: List<Home> get() = Collections.unmodifiableList(this.homes)

    override fun createNewHome(name: String, dimension: RegistryKey<World>, pos: BlockPos): Home {
        if (homesByName.containsKey(name)) throw HOME_EXISTS.create(name)
        val maxHomes = homesConfig!!.maxHomeAmount
        if (maxHomes != -1) {
            if (homesConfig!!.maxHomesPerDimension) {
                if (homesByWorld.getOrDefault(dimension.value, emptyList()).size >= maxHomes) {
                    throw TOO_MANY_HOMES.create(maxHomes)
                }
            } else {
                if (homes.size >= maxHomes) {
                    throw TOO_MANY_HOMES.create(maxHomes)
                }
            }
        }
        val home = Home(name, dimension, pos)
        homes.add(home)
        // Update maps for fast lookup.
        homesByName[home.name] = home
        homesByWorld.computeIfAbsent(home.dimension.value) { ArrayList() }.add(home)
        return home
    }

    override fun removeHome(home: Home) {
        homesByName.remove(home.name)
        homesByWorld.remove(home.dimension.value)
        homes.remove(home)
    }

    override fun removeHomeByName(name: String) {
        if (!homesByName.containsKey(name)) {
            throw HOME_NOT_FOUND.create(name)
        }
        removeHome(homesByName[name]!!)
    }

    override fun removeHomesInDimension(dimension: RegistryKey<World>) {
        homesByWorld.computeIfAbsent(dimension.value) { ArrayList() }.forEach { home ->
            removeHome(home)
        }
    }

    override fun readFromNbt(tag: NbtCompound) {
        homes.clear()
        homesByWorld.clear()
        homesByName.clear()
        val homes = tag.getList("Homes", 10)
        homes.stream().map { t -> t as NbtCompound }.map(Home::read).forEach { home -> this.homes.add(home) }

        // Compute the maps for easier lookup
        this.homes.forEach(Consumer { h: Home ->
            homesByName[h.name] = h
        })
        this.homes.forEach(Consumer { h: Home ->
            homesByWorld.computeIfAbsent(h.dimension.value) { ArrayList() }.add(h)
        })
    }

    override fun writeToNbt(tag: NbtCompound) {
        val homes = NbtList()
        this.homes.stream().map { home: Home ->
            home.write(
                NbtCompound()
            )
        }.forEach { e: NbtCompound -> homes.add(e) }
        tag.put("Homes", homes)
    }

    companion object {
        private val HOME_EXISTS = DynamicCommandExceptionType { name ->
            TranslatableText("serverutils.home.create.failed.exists", name)
        }
        val HOME_NOT_FOUND = DynamicCommandExceptionType { name ->
            TranslatableText("serverutils.home.tp.denied.notexists", name)
        }
        private val TOO_MANY_HOMES = DynamicCommandExceptionType { max ->
            TranslatableText("serverutils.home.create.failed.limit", max)
        }
    }
}